src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -133,24 +101,84 @@ static int copy(byte[] dest, int pos, String toCopy) {
         if (replyTo != null) {
             this.replyTo = replyTo;
         }
-        this.sizeInBytes = protocolLength + 2;
+        this.protocolLength = protocolLength;
         this.data = null; // will set data and size after we read it
     }
 
+    private static int fastUtf8Length(CharSequence cs) {
+        return cs.length() + cs
+                .codePoints()
+                .filter(cp -> cp>0x7f)
+                .map(cp -> cp<=0x7ff? 1: 2)
+                .sum();
+    }
+
+    private static int fastIntLength(int number) {
+        if (number < 100000) {
+            if (number < 100) {
+                if (number < 10) {
+                    return 1;
+                } else {
+                    return 2;
+                }
+            } else {
+                if (number < 1000) {
+                    return 3;
+                } else {
+                    if (number < 10000) {
+                        return 4;
+                    } else {
+                        return 5;
+                    }
+                }
+            }
+        } else {
+            if (number < 10000000) {
+                if (number < 1000000) {
+                    return 6;
+                } else {
+                    return 7;
+                }
+            } else {
+                if (number < 100000000) {
+                    return 8;
+                } else {
+                    if (number < 1000000000) {
+                        return 9;
+                    } else {
+                        return 10;
+                    }
+                }
+            }
+        }
+    }
+
     boolean isProtocol() {
         return this.subject == null;
     }
 
     // Will be null on an incoming message
     byte[] getProtocolBytes() {
-        return this.protocolBytes;
+        return this.protocolBytes.array();

```

Comment:
Since all `ByteBuffer`'s now have correctly allocated sizes we can unconditionally use the fast path here.